tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,216 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2019-2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "execute.hpp"
+#include "limits.hpp"
+#include "parser.hpp"
+#include <uvwasi.h>
+#include <cassert>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+
+namespace
+{
+// Global state.
+// NOTE: we do not care about uvwasi_destroy(), because it only clears file descriptors (currently)
+// and we are a single-run tool. This may change in the future and should reevaluate.
+uvwasi_t state;
+
+fizzy::ExecutionResult wasi_return_enosys(fizzy::Instance&, fizzy::span<const fizzy::Value>, int)
+{
+    return fizzy::Value{uint32_t{UVWASI_ENOSYS}};
+}
+
+fizzy::ExecutionResult wasi_proc_exit(fizzy::Instance&, fizzy::span<const fizzy::Value> args, int)
+{
+    uvwasi_proc_exit(&state, static_cast<uvwasi_exitcode_t>(args[0].as<uint32_t>()));
+    // Should not reach this.
+    return fizzy::Trap;
+}
+
+fizzy::ExecutionResult wasi_fd_write(
+    fizzy::Instance& instance, fizzy::span<const fizzy::Value> args, int)
+{
+    const auto fd = args[0].as<uint32_t>();
+    const auto iov_ptr = args[1].as<uint32_t>();
+    const auto iov_cnt = args[2].as<uint32_t>();
+    const auto nwritten_ptr = args[3].as<uint32_t>();
+
+    std::vector<uvwasi_ciovec_t> iovs(iov_cnt);
+    // TODO: not sure what to pass as end, passing memory size...
+    uvwasi_errno_t ret = uvwasi_serdes_readv_ciovec_t(instance.memory->data(),
+        instance.memory->size(), iov_ptr, iovs.data(), static_cast<uvwasi_size_t>(iovs.size()));
```

Comment:
I suggested this before already
```suggestion
        instance.memory->size(), iov_ptr, iovs.data(), iov_cnt);
```